### ISSUE TITLE
Improve performance when inspecting `@Nested` properties

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#24140]
As a Gradle plugin author, I don't want my non-trivial `@Nested` properties exempt from task dependencies to slow down task graph building, so the Gradle user's builds configuration stays optimal.

### Expected Behavior

Non-trivial task properties should not cause performance bottlenecks during task graph building (see Context).

### Current Behavior

`@Nested` properties are inspected in a very inefficient way creating a performance bottleneck during task graph building (see Context).

### Context

We understand the current behaviour for `@Nested` properties. A property annotated with `@Nested` is required to be inspected because any sub-properties may provide dependency information. In a way, it behaves similarly to `Provider#flatMap`. In my particular case, the `@Nested` is applied to a `Provider`. It's unclear if Gradle considers the implicit dependencies attached to the `Provider`, but it will be required to query the provider so it can inspect the nested properties. However, the computation of the value is not trivial, and Gradle's task graph building is executed serially, thus creating a performance bottleneck.

With that in mind, it seems what we need is either a way for Gradle to be more efficient when building the task graph (aka inspect properties in parallel) or a way to tell Gradle "trust me, there are no task dependencies in this nested object". Parallel inspection would be the best solution as the intended behaviour would stay the same, and no performance bottleneck would appear. Unfortunately, it would be a breaking change and potentially have multiple edge cases.

The current workaround is to inspect the stack trace to return a "no-op" value when task dependencies are computed. This is very evil but saves two whole seconds off the configuration phase (see [before](https://scans.gradle.com/s/kfhoe7ykt54ge/performance/build#build-time-configuration) and [after](https://scans.gradle.com/s/v4rxk77r3xa5o/performance/build#build-time-configuration)).

**Comment by lptr** [comment]
Hi @lacasseio, can you explain why the calculation of the `@Nested` value is not trivial in your use case? Our expectation is that `@Nested` properties should return the same pre-calculated instance in most use cases.

**Comment by lacasseio** [comment]
It's part of the Xcode build adapter where the property requires further parsing of the Xcode projects to determine the value. We want the value to be considered as part of the task snapshot and it doesn't provide any task dependencies. However, during the task graph calculation, the property will be inspected serially with the other tasks causing a bottleneck. Given they don't provide any task dependencies, the inspection is useless and only affects the performance. Moving the inspection to the execution phase, the task property could be inspected in parallel with other tasks but I believe is not done except under a configuration cache execution. So this also creates a bottleneck.

Due to the slow response on this issue, we worked around the initial problem by flattening the nested property into a Map of Object and marking that property as Input instead. The provider of the property correctly informs the task graph calculation code that no dependencies are present. However, we are still left with the second bottleneck with regards to parallel snapshot calculation aka marking a task as fully parallel from input/output snapshotting to the worker action execution. I believe there is another issue regarding this second performance bottleneck until the stale bot already closed it.
